Lib/test/test_coroutines.py

```diff
@@ -1224,17 +1224,13 @@ def __aexit__(self):
 
     def test_with_4(self):
         class CM:
-            def __enter__(self):
-                pass
-
-            def __exit__(self):
-                pass
+            pass
 
         async def foo():
             async with CM():
                 pass
 
-        with self.assertRaisesRegex(AttributeError, '__aexit__'):
+        with self.assertRaisesRegex(AttributeError, '__aenter__'):
```

Comment:
In addition to this test change, it would be good to add another test case that ensures the `__aexit__` attribute error is raised *before* the body of the with statement runs.

For example:

    class CM:
        async def __aenter__(self):
            return self

    body_executed = False
    async def foo():
        async with CM():
            body_executed = True

        # assertRaisesRegex for __aexit__
        self.assertFalse(body_executed)

That will also ensure the refcounting for that case is tested.

Reply:
Okay @ncoghlan, but I am not sure we have to define a new test case for this. I have added your `body_executed` assertion to the tests `test_with_2`, `test_with_3` and `test_with_4` instead, since they all raise an `AttributeError` (for the three cases only `__aenter__` defined, only `__aexit__` defined, and no method defined). What do you think?

Reply:
Aye, that's fine - I was only looking at the diff, and missed the existing test case.